server/src/main/java/com/cloud/vm/snapshot/VMSnapshotManagerImpl.java

```diff
@@ -358,9 +363,33 @@ public VMSnapshot allocVMSnapshot(Long vmId, String vsDisplayName, String vsDesc
             throw new InvalidParameterValueException("Can not snapshot memory when VM is not in Running state");
         }
 
+        List<VolumeVO> rootVolumes = _volumeDao.findReadyRootVolumesByInstance(userVmVo.getId());
+        if (rootVolumes == null || rootVolumes.isEmpty()) {
+            throw new CloudRuntimeException("Unable to find root volume for the user vm:" + userVmVo.getUuid());
+        }
+
+        VolumeVO rootVolume = rootVolumes.get(0);
+        StoragePoolVO rootVolumePool = _storagePoolDao.findById(rootVolume.getPoolId());
+        if (rootVolumePool == null) {
+            throw new CloudRuntimeException("Unable to find root volume storage pool for the user vm:" + userVmVo.getUuid());
+        }
+
         // for KVM, only allow snapshot with memory when VM is in running state
-        if (userVmVo.getHypervisorType() == HypervisorType.KVM && userVmVo.getState() == State.Running && !snapshotMemory) {
-            throw new InvalidParameterValueException("KVM VM does not allow to take a disk-only snapshot when VM is in running state");
+        if (userVmVo.getHypervisorType() == HypervisorType.KVM) {
+            if (rootVolumePool.getPoolType() != Storage.StoragePoolType.PowerFlex) {
```

Comment:
Hi @sureshanaparti, may I suggest here to get the VMSnapshotStrategy instead of checking which is the storage pool type? If it's an instance of DefaultVMSnapshotStrategy to get into the statement. In this case, the other storage plugins could implement their own VM snapshot strategy, without checking here which is the storage pool type. I saw that Solidfire's API has a group snapshot (maybe and others have it). We at StorPool also have this functionality (but for now we have a workaround for this). I guess someday it will be easier for the plugins to adopt this functionality

Reply:
@slavkap Thanks for the review and suggestions. Can not get the VMSnapshotStrategy to use here, as VMSnapshot object is not yet created. The check here is a pre-creation phase check, and  VMSnapshot object is not created when it is not supported. VMSnapshotStrategy is currently being used to check (_canHandle(VMSnapshot)_) and perform the operations (take VM snapshot, revert VM snapshots, delete VM snapshot) allowed, on the VMSnapshot object.

Reply:
Thanks, @sureshanaparti, for the reply! My bad that I didn't look below in the code, where the VMSnapshot object is created. My suggestion about this functionality was to be more general for the rest of the plugins if somebody wants to integrate with it. I will try to find a solution in our PR